DigitalLearningSolutions.Data/Services/DelegateApprovalsService.cs

```diff
@@ -1,34 +1,133 @@
     using DigitalLearningSolutions.Data.Models.CustomPrompts;
+    using DigitalLearningSolutions.Data.Models.Email;
     using DigitalLearningSolutions.Data.Models.User;
+    using Microsoft.Extensions.Logging;
+    using MimeKit;
 
     public interface IDelegateApprovalsService
     {
         public List<(DelegateUser delegateUser, List<CustomPromptWithAnswer> prompts)>
             GetUnapprovedDelegatesWithCustomPromptAnswersForCentre(int centreId);
+
+        public void ApproveDelegate(int delegateId, int centreId);
+        public void ApproveAllUnapprovedDelegatesForCentre(int centreId);
     }
 
     public class DelegateApprovalsService : IDelegateApprovalsService
     {
-        private readonly IUserDataService userDataService;
         private readonly ICustomPromptsService customPromptsService;
+        private readonly IEmailService emailService;
+        private readonly ILogger<DelegateApprovalsService> logger;
+        private readonly IUserDataService userDataService;
 
-        public DelegateApprovalsService(IUserDataService userDataService, ICustomPromptsService customPromptsService)
+        public DelegateApprovalsService
+        (
+            IUserDataService userDataService,
+            ICustomPromptsService customPromptsService,
+            IEmailService emailService,
+            ILogger<DelegateApprovalsService> logger
+        )
         {
             this.userDataService = userDataService;
             this.customPromptsService = customPromptsService;
+            this.emailService = emailService;
+            this.logger = logger;
         }
 
-        public List<(DelegateUser delegateUser, List<CustomPromptWithAnswer> prompts)> GetUnapprovedDelegatesWithCustomPromptAnswersForCentre(int centreId)
+        public List<(DelegateUser delegateUser, List<CustomPromptWithAnswer> prompts)>
+            GetUnapprovedDelegatesWithCustomPromptAnswersForCentre(int centreId)
         {
             var users = userDataService.GetUnapprovedDelegateUsersByCentreId(centreId);
             var usersWithPrompts =
                 customPromptsService.GetCentreCustomPromptsWithAnswersByCentreIdForDelegateUsers(centreId, users);
 
             return usersWithPrompts;
         }
+
+        public void ApproveDelegate(int delegateId, int centreId)
+        {
+            var delegateUser = userDataService.GetDelegateUserById(delegateId);
+
+            if (delegateUser == null || delegateUser.CentreId != centreId)
+            {
+                throw new UserAccountNotFoundException($"Delegate user id {delegateId} not found at centre id {centreId}.");
+            }
+
+            if (delegateUser.Approved)
+            {
+                logger.LogWarning($"Delegate user id {delegateId} already approved.");
+            }
+            else
+            {
+                var delegateList = new List<DelegateUser>{ delegateUser };
+                userDataService.ApproveDelegateUsers(delegateList.Select(d => d.Id));

```

Comment:
Could this just use the delegateId (or delegateUser.Id) as a parameter?

Reply:
I think you're right - I can just use `params` myself